docs/website/docs/general-usage/credentials/configuration.md

```diff
@@ -359,129 +359,5 @@ Lookup rules:
 **Rule 1:** The lookup starts with the most specific possible path, and if value is not found there,
 it removes the right-most section and tries again.
 
-Example: We use the `bigquery` destination and the `google_sheets` source. They both use google credentials and expect them to be configured under `credentials` key.
+Example: We use the `bigquery` destination and the `google_sheets` source. They both use
 
```

Comment:
```suggestion
 google credentials and expect them to be configured under `credentials` key.

1. If we create just a single `credentials` section like in [here](#default-layout-without-sections), destination and source will share the same credentials.

2. If we define sections as below, we'll keep the credentials separate

```toml
# google sheet credentials
[sources.credentials]
client_email = <client_email from services.json>
private_key = <private_key from services.json>
project_id = <project_id from services json>

# bigquery credentials
[destination.credentials]
client_email = <client_email from services.json>
private_key = <private_key from services.json>
project_id = <project_id from services json>
```

Now when `dlt` looks for destination credentials, it will start with `destination.bigquery.credentials`, eliminate `bigquery` and stop at `destination.credentials`.

When looking for `sources` credentials it will start with `sources.google_sheets.google_sheets.credentials`, eliminate `google_sheets` twice and stop at `sources.credentials` (we assume that `google_sheets` source was defined in `google_sheets` python module)

Example: let's be even more explicit and use a full section path possible.

```toml
# google sheet credentials
[sources.google_sheets.credentials]
client_email = <client_email from services.json>
private_key = <private_key from services.json>
project_id = <project_id from services json>

# google analytics credentials
[sources.google_analytics.credentials]
client_email = <client_email from services.json>
private_key = <private_key from services.json>
project_id = <project_id from services json>

# bigquery credentials
[destination.bigquery.credentials]
client_email = <client_email from services.json>
private_key = <private_key from services.json>
project_id = <project_id from services json>
```

Now we can separate credentials for different sources as well.

**Rule 2:** You can use your pipeline name to have separate configurations for each pipeline in your
project.

Pipeline created/obtained with `dlt.pipeline()` creates a global and optional namespace with the
value of `pipeline_name`. All config values will be looked with pipeline name first and then again
without it.

Example: the pipeline is named `ML_sheets`.

```toml
[ML_sheets.credentials]
client_email = <client_email from services.json>
private_key = <private_key from services.json>
project_id = <project_id from services json>
```

or maximum path:

```toml
[ML_sheets.sources.google_sheets.credentials]
client_email = <client_email from services.json>
private_key = <private_key from services.json>
project_id = <project_id from services json>
```

### The `sources` section

Config and secrets for decorated sources and resources are kept in
`sources.<source module name>.<function_name>` section. **All sections are optional during lookup**. For example,
if source module is named `pipedrive` and the function decorated with `@dlt.source` is
`deals(api_key: str=...)` then `dlt` will look for API key in:

1. `sources.pipedrive.deals.api_key`
1. `sources.pipedrive.api_key`
1. `sources.api_key`
1. `api_key`

Step 2 in a search path allows all the sources/resources in a module to share the same set of
credentials.

Also look at the [following test](https://github.com/dlt-hub/dlt/blob/devel/tests/extract/test_decorators.py#L303) `test_source_sections`.

## Understanding the exceptions

Now we can finally understand the `ConfigFieldMissingException`.

Let's run `chess.py` example without providing the password:

```
$ CREDENTIALS="postgres://loader@localhost:5432/dlt_data" python chess.py
...
dlt.common.configuration.exceptions.ConfigFieldMissingException: Following fields are missing: ['password'] in configuration with spec PostgresCredentials
        for field "password" config providers and keys were tried in following order:
                In Environment Variables key CHESS_GAMES__DESTINATION__POSTGRES__CREDENTIALS__PASSWORD was not found.
                In Environment Variables key CHESS_GAMES__DESTINATION__CREDENTIALS__PASSWORD was not found.
                In Environment Variables key CHESS_GAMES__CREDENTIALS__PASSWORD was not found.
                In secrets.toml key chess_games.destination.postgres.credentials.password was not found.
                In secrets.toml key chess_games.destination.credentials.password was not found.
                In secrets.toml key chess_games.credentials.password was not found.
                In Environment Variables key DESTINATION__POSTGRES__CREDENTIALS__PASSWORD was not found.
                In Environment Variables key DESTINATION__CREDENTIALS__PASSWORD was not found.
                In Environment Variables key CREDENTIALS__PASSWORD was not found.
                In secrets.toml key destination.postgres.credentials.password was not found.
                In secrets.toml key destination.credentials.password was not found.
                In secrets.toml key credentials.password was not found.
Please refer to https://dlthub.com/docs/general-usage/credentials for more information
```

It tells you exactly which paths `dlt` looked at, via which config providers and in which order.

In the example above:

1. First it looked in a big section `chess_games` which is name of the pipeline.
1. In each case it starts with full paths and goes to minimum path `credentials.password`.
1. First it looks into `environ` then in `secrets.toml`. It displays the exact keys tried.
1. Note that `config.toml` was skipped! It may not contain any secrets.

Read more about [Provider Hierarchy](./config_providers).
```